pkgs/api/src/lib.zig

```diff
@@ -58,33 +76,60 @@ pub const Timer = struct {
         // For freestanding targets where we can't measure time, just record 0
         const duration_seconds = if (duration_ns == 0) 0.0 else @as(f32, @floatFromInt(duration_ns)) / 1_000_000_000.0;
 
-        if (self.is_chain) {
-            metrics.chain_onblock_duration_seconds.observe(duration_seconds);
-        } else {
-            metrics.block_processing_duration_seconds.observe(duration_seconds);
-        }
+        self.observeFn(self.context, duration_seconds);
 
         return duration_seconds;
     }
 };
 
 /// A wrapper struct that exposes a `start` function to match the existing API.
 pub const Histogram = struct {
-    is_chain: bool,
+    context: *anyopaque,
+    observeFn: *const fn (*anyopaque, f32) void,
 
     pub fn start(self: *const Histogram) Timer {
         return Timer{
             .start_time = getTimestamp(),
-            .histogram = undefined, // Not used in this implementation
-            .is_chain = self.is_chain,
+            .context = self.context,
+            .observeFn = self.observeFn,
         };
     }
 };
 
+// Type-erased observe functions for each histogram type
+fn observeChainOnblock(ctx: *anyopaque, value: f32) void {
+    const histogram: *Metrics.ChainHistogram = @ptrCast(@alignCast(ctx));
+    histogram.observe(value);
+}
+
+fn observeBlockProcessing(ctx: *anyopaque, value: f32) void {
+    const histogram: *Metrics.BlockProcessingHistogram = @ptrCast(@alignCast(ctx));
+    histogram.observe(value);
+}
+
+fn observeStateTransition(ctx: *anyopaque, value: f32) void {
+    const histogram: *Metrics.StateTransitionHistogram = @ptrCast(@alignCast(ctx));
+    histogram.observe(value);
+}
+
+// No-op observe function for when metrics are not initialized
+fn observeNoop(ctx: *anyopaque, value: f32) void {
+    _ = ctx;
+    _ = value;
+}
+
 /// The public variables the application interacts with.
 /// Calling `.start()` on these will start a new timer.
-pub var chain_onblock_duration_seconds: Histogram = Histogram{ .is_chain = true };
-pub var block_processing_duration_seconds: Histogram = Histogram{ .is_chain = false };
+/// Initialized with no-op defaults; properly initialized in init() after the metrics are created.
+/// This allows them to be used safely even if init() is never called (e.g., in tests).
+pub var chain_onblock_duration_seconds: Histogram = .{
+    .context = undefined,
+    .observeFn = &observeNoop,
+};
+pub var block_processing_duration_seconds: Histogram = .{
+    .context = undefined,
+    .observeFn = &observeNoop,
+};
```

Comment:
I see, you are forced to use function pointers based on the noop. But then I suggest you do use a `?Histogram` and only report if it's not `null`